### ISSUE TITLE
VictoriaMetrics breaks on electricity loss

### ISSUE THREAD

**Issue by erkexzcx** [VictoriaMetrics/VictoriaMetrics#5012]
### Describe the bug

I am hosting VictoriaMetrics as Docker container on my intel nuc at home. Storage is NFS mountpoint (managed by Docker) on my Synology NAS:
```yaml
services:
  victoriametrics:
    container_name: victoriametrics
    image: victoriametrics/victoria-metrics:stable
    restart: always
    volumes:
      - victoriametrics:/storage
    command:
      - "--storageDataPath=/storage"

volumes:
  victoriametrics:
    driver_opts:
      type: "nfs4"
      o: "addr=123.123.123.123,rsize=32768,wsize=32768,hard,timeo=600,retrans=2,noresvport,nconnect=16,vers=4.1"
      device: ":/volume1/victoriametrics"
```

And I live in a somewhat remote location where power loss is usually around once a month or two. My intel nuc has no UPS or battery, so it simply shutdowns together with the NAS and network equipment.

Usually everything starts automatically and just works when power is restored, but this is 2nd time when this happens - VictoriaMetrics does not get up:
```
        github.com/VictoriaMetrics/VictoriaMetrics/app/victoria-metrics/main.go:74 +0x325
2023-09-15T09:20:37.455Z        info    VictoriaMetrics/lib/logger/flag.go:12   build version: victoria-metrics-20230812-215739-tags-v1.93.0-0-g59f7d810c9
2023-09-15T09:20:37.455Z        info    VictoriaMetrics/lib/logger/flag.go:13   command-line flags
2023-09-15T09:20:37.455Z        info    VictoriaMetrics/lib/logger/flag.go:20     -storageDataPath="/storage"
2023-09-15T09:20:37.455Z        info    VictoriaMetrics/app/victoria-metrics/main.go:70 starting VictoriaMetrics at ":8428"...
2023-09-15T09:20:37.455Z        info    VictoriaMetrics/app/vmstorage/main.go:108       opening storage at "/storage" with -retentionPeriod=1
2023-09-15T09:20:37.457Z        info    VictoriaMetrics/lib/memory/memory.go:42 limiting caches to 20011091558 bytes, leaving 13340727706 bytes to the OS according to -memory.allowedPercent=60
2023-09-15T09:20:37.480Z        info    VictoriaMetrics/lib/storage/storage.go:935      discarding /storage/cache/curr_hour_metric_ids, since it contains outdated hour; got 470638; want 470769
2023-09-15T09:20:37.480Z        info    VictoriaMetrics/lib/storage/storage.go:935      discarding /storage/cache/prev_hour_metric_ids, since it contains outdated hour; got 470637; want 470768
2023-09-15T09:20:37.497Z        info    VictoriaMetrics/lib/storage/storage.go:896      discarding /storage/cache/next_day_metric_ids_v2, since it contains data for stale date; got 19609; want 19615
2023-09-15T09:20:37.554Z        panic   VictoriaMetrics/lib/storage/part_header.go:150  FATAL: cannot parse "/storage/data/small/2023_09/17835E2B9C49FBA7/metadata.json": unexpected end of JSON input
panic: FATAL: cannot parse "/storage/data/small/2023_09/17835E2B9C49FBA7/metadata.json": unexpected end of JSON input

goroutine 1 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logMessage({0xbc255d, 0x5}, {0xc00008f3b0, 0x6e}, 0x2?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:281 +0xa91
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevelSkipframes(0x1, {0xbc255d, 0x5}, {0xbd1159?, 0xc000388f60?}, {0xc0004dac70?, 0x0?, 0xc0004dabf8?})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:138 +0x194
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevel(...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:130
github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.Panicf(...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:126
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*partHeader).MustReadMetadata(0xc000388f60, {0xc000388f30, 0x2c})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part_header.go:150 +0x26c
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenFilePart({0xc000388f30?, 0x1b?})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:53 +0x65
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenParts({0xc000342220, 0x1b}, {0xc000386000, 0x16, 0x4ca0a0?})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:1805 +0x433
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenPartition({0xc000342220?, 0xc0004db258?}, {0xc000342240, 0x19}, 0xc00069f140?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:267 +0x246
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenPartitions({0xc00002a690, 0x13}, {0xc00002a7c8, 0x11}, 0x0?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/table.go:480 +0x275
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenTable({0xc0006b0080?, 0x1781cdc4ed5ead07?}, 0xc000269a00)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/table.go:103 +0x269
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.MustOpenStorage({0x7ffd4044bf75?, 0xbc1ff6?}, 0x9fa52400, 0x0, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/storage.go:275 +0xfe5
github.com/VictoriaMetrics/VictoriaMetrics/app/vmstorage.Init(0xd77bf0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmstorage/main.go:111 +0x505
main.main()
        github.com/VictoriaMetrics/VictoriaMetrics/app/victoria-metrics/main.go:74 +0x325
```

Well, this is home lab and the data is not really critical - I can just delete everything and restart it, but I am raising this issue to bring awareness about unclean shutdowns and the way they break VictoriaMetrics.

Anything to suggest? Or can this issue be closed?

### To Reproduce

1. Host VictoriaMetrics (non cluster version), where data is mounted as NFS share via Docker.
2. Setup some other hosts with Telegraf and write to VictoriaMetrics (as influxdb host, via HTTP port) every 10 seconds.
3. Switch off electricity
4. Switch on electricity
5. Everything boots, but VictoriaMetrics fails with given error logs.

### Version

Docker image: `victoriametrics/victoria-metrics:stable` (`victoriametrics/victoria-metrics:v1.93.0`)
Binaries in Docker image: `victoria-metrics-20230812-215739-tags-v1.93.0-0-g59f7d810c9`

### Logs

_Pasted above_

### Screenshots

N/A

### Used command-line flags

_Pasted above_

### Additional information

Everything explained above

**Comment by f41gh7** [comment]
Hello, VictoriaMetrics uses `fsync` to ensure, that data was written into filesystem. After write acknowledge, it updates value of parts.json file. 
  In your case, data was cached into driver buffer and completely lost due to electricity outage. 

 I suggest: 
1) shutdown VictoriaMetrics
2)  manually edit `parts.json` file - removing deleted part names. It's located at `/data/PART/small directory. And contains both `small` and `big` part names.
3) start VictoriaMetrics and check logs for any errors.

**Comment by valyala** [comment]
FYI, [the fix](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5017) has been included in [v1.94.0 release](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by hagen1778** [comment]
@erkexzcx could you please show the content of metdata.json? Is it true that it has 0 bytes and is empty? 
If yes, you should remove files with 0 bytes size.

**Comment by erkexzcx** [comment]
Yup, it's empty:
```
root@SynologyNAS:/volume1/victoriametrics/data/small/2023_09/17835E2B9C49FBA8# ll
total 0
d---------+ 1 root root 118 Sep 14 21:44 .
d---------+ 1 root root 692 Sep 20 10:42 ..
----------+ 1 root root   0 Sep 14 21:44 index.bin
----------+ 1 root root   0 Sep 14 21:44 metadata.json
----------+ 1 root root   0 Sep 14 21:44 metaindex.bin
----------+ 1 root root   0 Sep 14 21:44 timestamps.bin
----------+ 1 root root   0 Sep 14 21:44 values.bin
root@SynologyNAS:/volume1/victoriametrics/data/small/2023_09/17835E2B9C49FBA8# cat *
root@SynologyNAS:/volume1/victoriametrics/data/small/2023_09/17835E2B9C49FBA8# 
```

Let's remove `metadata.json`:
```
2023-11-17T21:30:36.633Z        panic   VictoriaMetrics/lib/storage/part_header.go:142  FATAL: cannot parse metadata from "/storage/data/small/2023_09/17835E2B9C49FBA8": unexpected number of substrings in the part name "17835E2B9C49FBA8": got 1; want 5
panic: FATAL: cannot parse metadata from "/storage/data/small/2023_09/17835E2B9C49FBA8": unexpected number of substrings in the part name "17835E2B9C49FBA8": got 1; want 5
```

Let's remove all files from `/storage/data/small/2023_09/17835E2B9C49FBA8` - same error.

Now let's remove a whole folder `/volume1/victoriametrics/data/small/2023_09/17835E2B9C49FBA8`:
```
2023-11-17T21:32:34.572Z        panic   VictoriaMetrics/lib/storage/partition.go:1772   FATAL: part "/storage/data/small/2023_09/17835E2B9C49FBA8" is listed in "/storage/data/small/2023_09/parts.json", but is missing on disk; ensure "/storage/data/small/2023_09/parts.json" contents is not corrupted; remove "/storage/data/small/2023_09/parts.json" to rebuild its' content from the list of existing parts
panic: FATAL: part "/storage/data/small/2023_09/17835E2B9C49FBA8" is listed in "/storage/data/small/2023_09/parts.json", but is missing on disk; ensure "/storage/data/small/2023_09/parts.json" contents is not corrupted; remove "/storage/data/small/2023_09/parts.json" to rebuild its' content from the list of existing parts
```

Anyway, so let's remove `17835E2B9C49FBA8` from `/storage/data/small/2023_09/parts.json`:
```
2023-11-17T21:33:40.688Z        panic   VictoriaMetrics/lib/storage/part_header.go:150  FATAL: cannot parse "/storage/data/small/2023_09/17835E2B9C49FBA9/metadata.json": unexpected end of JSON input
panic: FATAL: cannot parse "/storage/data/small/2023_09/17835E2B9C49FBA9/metadata.json": unexpected end of JSON input
```

Oh no! Another folder having issues:
```
root@SynologyNAS:/volume1/victoriametrics/data/small/2023_09# ls -l 17835E2B9C49FBA9
total 0
----------+ 1 root root 0 Sep 14 21:44 index.bin
----------+ 1 root root 0 Sep 14 21:44 metadata.json
----------+ 1 root root 0 Sep 14 21:44 metaindex.bin
----------+ 1 root root 0 Sep 14 21:44 timestamps.bin
----------+ 1 root root 0 Sep 14 21:44 values.bin
root@SynologyNAS:/volume1/victoriametrics/data/small/2023_09#
```

Ain't my lucky day:
```
root@SynologyNAS:/volume1/victoriametrics/data/small/2023_09# du -sh *
11M     1781CDC51EDB011A
9.3M    1781CDC51EDB4F92
9.8M    1782DB600B409348
8.5M    17835E2B9C48E304
5.9M    17835E2B9C4920C6
5.6M    17835E2B9C4970F0
8.6M    17835E2B9C49B10F
1.1M    17835E2B9C49C22E
5.5M    17835E2B9C49EDB9
1.1M    17835E2B9C49F71C
180K    17835E2B9C49F7D6
188K    17835E2B9C49F891
216K    17835E2B9C49F989
212K    17835E2B9C49FA75
204K    17835E2B9C49FB48
84K     17835E2B9C49FB7B
40K     17835E2B9C49FB86
40K     17835E2B9C49FB93
40K     17835E2B9C49FB9D
0       17835E2B9C49FBA9
4.0K    parts.json
```

Let me remove the `17835E2B9C49FBA9` and modify `parts.json` accordingly... And :tada: - VictoriaMetrics has started! :)

It just feels like I should be writing a Python script and launch it before starting VictoriaMetrics, but I believe this has to be fixed from your side? AFAIK VictoriaLogs is also impacted by this bug :(

**Comment by hagen1778** [comment]
> It just feels like I should be writing a Python script and launch it before starting VictoriaMetrics, but I believe this has to be fixed from your side? AFAIK VictoriaLogs is also impacted by this bug :(

I don't think it needs to be fixed on the VictoriaMetrics side due to the following reasons:
1. VictoriaMetrics doesn't leave partially written files (like 0 byte files in your case):
> Newly added parts either successfully appear in the storage or fail to appear. The newly added part is atomically registered in the parts.json file under the corresponding partition after it is fully written and [fsynced](https://man7.org/linux/man-pages/man2/fsync.2.html) to the storage. Thanks to this algorithm, storage never contains partially created parts, even if hardware power off occurs in the middle of writing the part to disk - such incompletely written parts are automatically deleted on the next VictoriaMetrics start.
See more here https://docs.victoriametrics.com/#storage
2. When VictoriaMetrics creates a part, it does it a temporary dir. Only if all write and sync operations succeed in temporary dir - VM moves temp dir to /data path. This operation is atomic in POSIX file systems. There is no transitioning state here.
3. In your case, all files in the part are there but with 0 size. This is likely a filesystem fault during moving data from one location to another: files were created but not synced. 
4. VictoriaMetrics could delete everything it can't read, but that would only hide a problem. If data corruption happens not because of VictoriaMetrics fault - user needs to be notified about it. Just like in this case.

**Comment by Frankkkkk** [comment]
Hi @hagen1778 I partly agree, but a auto-healing of obviously wrong files would be nice to have too. On my side I have basically the same setup as @erkexzcx and had the same fault too. Maybe not VM's fault, but auto repair would be plenty nice in order not to have another elastic-like DRP scenario

**Comment by hagen1778** [comment]
>  Maybe not VM's fault, but auto repair would be plenty nice in order not to have another elastic-like DRP scenario

We can keep this issue open to see if it gets more traction. But my point is still here: auto-deleting corrupted files hides the root cause of corruption. If VictoriaMetrics does this auto-deletion and user gets data loss unnoticed - then it would be extremely hard to find why the data was lost.

**Comment by valyala** [comment]
VictoriaMetrics calls [fsync](https://en.m.wikipedia.org/wiki/Sync_(Unix)) on all the created files and directories in order to make sure the data and the directory structure is successfully written to the underlying storage. If `fsync` fails, then VictoriaMetrics logs the error and exits with non-zero exit code. It doesn't remove the source files during the background merge in this case, so the original data isn't lost after the restart. It automatically drops incompletely stored parts on the next start, if these parts aren't mentioned in the `parts.json` file.

The `parts.json` file is updated atomically after all the parts data is successfully stored and fsync'ed to disk. The atomic update of this file is executed in the following sequence:

- A temporary file is created with the updates list of parts. This file is created in the same directory where the original `parts.json` file is located.
- The temporary file is atomically moved to `parts.json` with [`rename`](https://man7.org/linux/man-pages/man2/rename.2.html). This syscall is guaranteed to be atomic:

> If newpath already exists, it will be atomically replaced, so
       that there is no point at which another process attempting to
       access newpath will find it missing.  However, there will
       probably be a window in which both oldpath and newpath refer to
       the file being renamed.

I suspect that the NFS implementation you use may have the following issues:

- It doesn't guarantee that the data is successfully written to the remote storage after returning from `fsync` syscall.
- It doesn't guarantee that the file is successfully renamed at the remote storage after returning from `rename` syscall.

Both issues may result in partial data loss and crashes you experience. I still think that there are workarounds exist for these issues - we just need to identify the real cause and then figure out the workaround. Probably it is related to NFS mount options.

Probably related NFS issues:

- https://unix.stackexchange.com/questions/265341/nfs-commit-operation-not-being-invoked-despite-fsync-being-used
- https://stackoverflow.com/questions/41362016/rename-atomicity-and-nfs

**Comment by Frankkkkk** [comment]
I guess an option would be to mount the NFS share with the `sync` option. Will be slower, but it may fix the bug (lost some data again ^^').

**Comment by makasim** [comment]
I encountered the same (or a very similar) issue. Due to repeated power outages, my homelab ended up in this state. Both VM and VL reported similar problems. They are running on ext4 NVMe disks.

```
Mar 18 10:04:32.135 homelab vmsingle[b0355] 2026-03-18T08:04:32.135Z    panic   VictoriaMetrics/lib/storage/part_header.go:150  FATAL: cannot parse "/mnt/shared/victoriametrics-data/data/small/2026_02/1897F2658B667882/metadata.json": invalid character '`' looking for beginning of value
Mar 18 10:04:32.139 homelab vmsingle[b0355] panic: FATAL: cannot parse "/mnt/shared/victoriametrics-data/data/small/2026_02/1897F2658B667882/metadata.json": invalid character '`' looking for beginning of value [recovered, repanicked]
Mar 18 10:04:32.139 homelab vmsingle[b0355]
Mar 18 10:04:32.139 homelab vmsingle[b0355] goroutine 263 [running]:
Mar 18 10:04:32.139 homelab vmsingle[b0355] sync.(*WaitGroup).Go.func1.1()
Mar 18 10:04:32.139 homelab vmsingle[b0355]     sync/waitgroup.go:251 +0x48
Mar 18 10:04:32.139 homelab vmsingle[b0355] panic({0xbc12c0?, 0xd511848650?})
Mar 18 10:04:32.139 homelab vmsingle[b0355]     runtime/panic.go:860 +0x12c
Mar 18 10:04:32.139 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logMessageInternal({0xca6fc5, 0x5}, {0xd511960780, 0x9e}, {0xd5122243c0, 0x2e})
Mar 18 10:04:32.139 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:329 +0x6fc
Mar 18 10:04:32.139 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevelSkipframes(0xd511ccc8a0?, {0xca6fc5, 0x5}, {0xcbcc83, 0x1a}, {0xd511f05748, 0x2, 0x2})
Mar 18 10:04:32.139 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:162 +0x2d8
Mar 18 10:04:32.139 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevel(...)
Mar 18 10:04:32.139 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:147
Mar 18 10:04:32.139 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.Panicf(...)
Mar 18 10:04:32.139 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:143
Mar 18 10:04:32.139 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*partHeader).MustReadMetadata(0xd512224360, {0xd511da4fa0, 0x44})
Mar 18 10:04:32.139 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part_header.go:150 +0x198
Mar 18 10:04:32.139 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenFilePart({0xd511da4fa0?, 0xd5124d8840?})
Mar 18 10:04:32.139 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/part.go:53 +0x50
Mar 18 10:04:32.139 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenParts({0xd511847280, 0x3e}, {0xd511847080, 0x33}, {0xd511989600, 0x1a, 0x0?})
Mar 18 10:04:32.140 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:1986 +0x410
Mar 18 10:04:32.140 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenPartition({0xd511847080?, 0x0?}, {0xd5118470c0, 0x31}, {0xd511847100, 0x35}, 0xd511b42000)
Mar 18 10:04:32.140 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/partition.go:285 +0x360
Mar 18 10:04:32.140 homelab vmsingle[b0355] github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.mustOpenPartitions.func1()
Mar 18 10:04:32.140 homelab vmsingle[b0355]     github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/table.go:657 +0xf8
Mar 18 10:04:32.140 homelab vmsingle[b0355] sync.(*WaitGroup).Go.func1()
Mar 18 10:04:32.140 homelab vmsingle[b0355]     sync/waitgroup.go:258 +0x48
Mar 18 10:04:32.140 homelab vmsingle[b0355] created by sync.(*WaitGroup).Go in goroutine 1
Mar 18 10:04:32.140 homelab vmsingle[b0355]     sync/waitgroup.go:238 +0x70
```

metadata.json looks very suspicious: 
```
makasim@homelab:/mnt/shared/victoriametrics-data $ cat /mnt/shared/victoriametrics-data/data/small/2026_02/1897F2658B667882/metadata.json
CT𜎼�N   �N"���TS�
                  ���'��"vM�S
                              Sgx��Р(�v����Y*�O���m`M(`�fE�wk
Z�#�@��e�V������\-M£B�@��0g/
                             ΅���
```
